lib/protocol2.js

```diff
@@ -0,0 +1,254 @@
+    }
+    return this.closed ? null : this._sync(id)
+  }
+
+  _alloc () {
+    const id = this._byId.indexOf(null)
+    return id === -1 ? this._byId.push(null) - 1 : id
+  }
+
+  _setRemoteId (id, peer) {
+    // TODO: Some better abstraction for this that is safer, allow MAX backlog...
+    while (id >= this._byRemoteId.length) this._byRemoteId.push(null)
+    this._byRemoteId[id] = peer
+  }
+
+  _get (dk, upsert) {
+    const k = dk.toString('hex')
+
+    let map = this._byDiscoveryKey.get(k)
+    if (map || !upsert) return map
+
+    map = [0, 0]
+    this._byDiscoveryKey.set(k, map)
+    return map
+  }
+
+  getPeer (discoveryKey) {
+    const map = this._get(discoveryKey, false)
+    if (!map) return null
+    return this._byId[map[0]]
+  }
+
+  hasPeer (discoveryKey) {
+    return this.getPeer(discoveryKey) !== null
+  }
+
+  addPeer (discoveryKey, peer) {
+    if (this.closed) throw new Error('Protocol is closed')
+
+    const id = this._alloc()
+    const map = this._get(discoveryKey, true)
+
+    map[0] = id
+    this._byId[id] = peer
+
+    return id
+  }
+}
+
+async function onopen (msg, protocol) {
+  let map = protocol._get(msg.discoveryKey, false)
+
+  // Allow the receiver a chance to open the core if they want to
+  if (!map) {
+    const p = {
+      id: msg.id,
+      discoveryKey: msg.discoveryKey,
+      promise: null,
+      done: null
+    }
+
+    p.promise = new Promise((resolve) => { p.done = resolve })
+
+    protocol._opens.push(p)
+
+    try {
+      await protocol.onunknowncore(msg.discoveryKey)
+    } finally {
+      const i = protocol._opens.indexOf(p)
+      if (i > -1) protocol._opens.splice(i, 1)
+      p.done()
+    }
+  }
+
+  if (protocol.closed) return
+
+  map = protocol._get(msg.discoveryKey, false)
+
+  // Receiver didn't open it, send close signal
+  if (!map) {
+    protocol.close.send({ discoveryKey: msg.discoveryKey })
+    return
+  }
+
+  const peer = protocol._byId[map[0]]
+
+  // Check if remote already mapped it
+  if (map[1] < protocol._byRemoteId.length && peer === protocol._byRemoteId[map[1]]) {
+    throw new Error('Duplicate open')
+  }
+
+  // TODO: ... techically this should be async but, eh, too much work so sync for now
+  // could be done by pipelining the entire onopen func tho prob
+  peer.onremoteopen(msg)
+
+  protocol._setRemoteId(msg.id, peer)
+}
+
+async function onclose (msg, protocol) {
+  for (const p of protocol._opens) {
+    if (b4a.equals(p.discoveryKey, msg.discoveryKey)) await p.promise
+  }
+
+  if (protocol.closed) return
+
+  const map = protocol._get(msg.discoveryKey)
+  if (!map) return
+
+  const peer = protocol._byId[map[0]]
+
+  // unref the peer
+  protocol._byId[map[0]] = null
+  if (map[1] < protocol._byRemoteId.length && protocol._byRemoteId[map[1]] === peer) {
+    protocol._byRemoteId[map[1]] = null
+  }
+  protocol._byDiscoveryKey.delete(msg.discoveryKey.toString('hex'))
+
+  return peer.onremoteclose()
+}
+
+async function onsync (msg, protocol) {
```

Comment:
Would consider wrapping these in a pattern since the only relevant parts are `onrequest` -> `peer.onrequest(msg)` etc.

Reply:
pattern?

Reply:
Ah, I think I know. It's super hot code (prob the most hot in hypercore) so wanna help the vm but making static 'copy-pasted' functions so they are easy to inline - assuming that is what you mean